packages/app/src/cli/services/dev/app-events/app-event-watcher.ts

```diff
@@ -0,0 +1,279 @@
+  -> file-watcher will emit a `extension_folder_created` event
+    -> app-event-watcher will determine that an extension(s) was(were) created and reload the app
+      -> The consumer will receive the updated app and the created extension(s)
+
+3. A directory is removed from the file system (/extensions/my_extension)
+  -> file-watcher will emit a `extension_folder_deleted` event
+    -> app-event-watcher will determine that all extensions in `my_extension` were deleted and remove them from the app
+      -> The consumer will receive the updated app and the deleted extension(s)
+
+4. A toml file is updated (/extensions/my_extension/extension.toml)
+  -> file-watcher will emit a `toml_updated` event
+    -> app-event-watcher will compare the old and new extensions to determine which were created, deleted or updated
+      -> The consumer will receive the updated app and the created, deleted and updated extensions
+
+5. The app.toml is updated
+  -> file-watcher will emit a `app_config_updated` event
+    -> app-event-watcher will compare the old and new config to determine which extensions were created, deleted or updated
+      -> The consumer will receive the updated app and the created, deleted and updated extensions
+ */
+
+/**
+ * The type of the extension event
+ *
+ * - Updated: The extension was updated (a file changed, but is not a source file, so it won't require a rebuild)
+ * - UpdatedSourceFile: The extension was updated and a source file was changed, so it will require a rebuild
+ * - Deleted: The extension was deleted
+ * - Created: The extension was created
+ */
+export enum EventType {
+  Updated,
+  UpdatedSourceFile,
+  Deleted,
+  Created,
+}
+
+export interface ExtensionEvent {
+  type: EventType
+  extension: ExtensionInstance
+}
+
+/**
+ * An AppEvent is the result of processing a file system event.
+ * It includes the updated app and the affected extensions.
+ * The startTime is the time when the initial file-system event was received, it can be used by the consumer
+ * to determine how long it took to process the event.
+ */
+interface AppEvent {
+  app: AppInterface
+  extensionEvents: ExtensionEvent[]
+  path: string
+  startTime: [number, number]
+}
+
+interface HandlerInput {
+  event: WatcherEvent
+  app: AppInterface
+  extensions: ExtensionInstance[]
+  options: OutputContextOptions
+}
+
+type Handler = (input: HandlerInput) => Promise<AppEvent>
+
+const handlers: {[key in WatcherEvent['type']]: Handler} = {
+  extension_folder_deleted: ExtensionFolderDeletedHandler,
+  extension_folder_created: ExtensionFolderCreatedHandler,
+  file_created: FileChangeHandler,
+  file_deleted: FileChangeHandler,
+  file_updated: FileChangeHandler,
+  toml_updated: TomlChangeHandler,
+  app_config_updated: AppConfigUpdatedHandler,
+  app_config_deleted: AppConfigDeletedHandler,
+}
+
+/**
+ * App event watcher will emit events when changes are detected in the file system.
+ */
+export class AppEventWatcher extends EventEmitter {
```

Comment:
How does this all behave at speed? If I change lots and lots of files, do we sequentially reload the app lots of times? Are the events buffered (in memory), any debouncing happening here?

Reply:
the app is reloaded when you change a toml file or add a new extension.
I'm not debouncing changes here, so if you edit a toml file a lot of times, it will trigger a lot of reloads.

I think adding a debouncer here makes sense, but that's an improvement that can happen in a follow up PR to not delay this one more.